lib/inspec/config.rb

```diff
@@ -0,0 +1,286 @@
+# Represents InSpec configuration.  Merges defaults, config file options,
+# and CLI arguments.
+
+require 'pp'
+
+module Inspec
+  class Config
+    # These are options that apply to any transport
+    GENERIC_CREDENTIALS = %w{
+      backend
+      sudo
+      sudo_password
+      sudo_command
+      sudo_options
+      shell
+      shell_options
+      shell_command
+    }.freeze
+
+    extend Forwardable
+
+    # Many parts of InSpec expect to treat the Config as a Hash
+    def_delegators :@final_options, :each, :delete, :[], :[]=, :key?
+    attr_reader :final_options
+
+    def initialize(cli_opts = {}, cfg_io = nil, command_name = nil)
+      @command_name = command_name || (ARGV.empty? ? nil : ARGV[0].to_sym)
+      @defaults = Defaults.for_command(@command_name)
+
+      @cli_opts = cli_opts.dup
+      cfg_io = resolve_cfg_io(@cli_opts, cfg_io)
+      @cfg_file_contents = read_cfg_file_io(cfg_io)
+
+      @merged_options = merge_options
+      @final_options = finalize_options
+    end
+
+    def diagnose
+      return unless self[:diagnose]
+      puts "InSpec version: #{Inspec::VERSION}"
+      puts "Train version: #{Train::VERSION}"
+      puts 'Command line configuration:'
+      pp @cli_opts
+      puts 'JSON configuration file:'
+      pp @cfg_file_contents
+      puts 'Merged configuration:'
+      pp @merged_options
+      puts
+    end
+
+    #-----------------------------------------------------------------------#
+    #                         Reading Config Files
+    #-----------------------------------------------------------------------#
+
+    # Regardless of our situation, end up with a readable IO object
+    def resolve_cfg_io(cli_opts, cfg_io)
```

Comment:
We don't expect cfg_io to be a StringIO - it can be anything IO-like. In actual use that's likely to be an IO::File, while in testing it is easier to use a StringIO.  If there is no file, then we use a StringIO to treat it in the same way.

Reply:
Guard clauses are used to short-circuit logic, usually a break, next, or return.  Use them you have things like `return true if condition...`.

In this case, there is a lot of logic on either side of the unless.  That's not a situation for a guard clause.

Reply:
I don't guess I'm following, I don't see any logic surrounding unless, couldn't the first line be:

```
return cfg_io if cfg_io
```

Reply:
I do think we need some sort of type checking on `cfg_io` as well too, otherwise anything that doesn't respond to `.read` will fail here: 

https://github.com/inspec/inspec/blob/09e0579f5c34b642de939dbfb54aaf220249bf03/lib/inspec/config.rb#L76

Reply:
I personally don't see that (passing a non-IO object) as likely - this is an internal InSpec object - but it's no big deal to add a duck check.

Reply:
And regarding the guard clause: you are totally right, I just didn't read down far enough.  This is an ideal case for a guard clause.